lib/system.nim

```diff
@@ -2769,41 +2769,83 @@ template once*(body: untyped): untyped =
 
 {.pop.} # warning[GcMem]: off, warning[Uninit]: off
 
-proc substr*(s: openArray[char]): string =
-  ## Copies a slice of `s` into a new string and returns this new
-  ## string.
+template NotJSnotVMnotNims(): static bool = # hack, see: #12517 #12518
+  when nimvm:
+    false
+  else:
+    notJSnotNims
+
+proc substr*(a: openArray[char]): string =
+  ## Returns a new string, copying contents of `a`.
+  ##
+  ## .. warning:: No input validation is performed! This can fail raising an
```

Comment:
this looks off - I don't see why this should be a special case, ie you can perform the length check "manually" before the copyMem. Having these odd/random exceptions is hard to keep track of.

_technically_ you should also copy the "correct" part before the overflow check for 1:1 compat with the old code, though that feels like stretching it.

Reply:
To clarify, I didn't change the behaviour of any procs in this PR,  as this would require a discussion, only the implementation details. If you ask me, this proc should be called `toString`/`copy`/`toOwned` and all `substr`s should have the same semantics.

Reply:
we're corrected several such "missing defect" bugs in the past, I wouldn't see this one as different - it's pretty clear cut that there should be a defect on oob, the only question is when (ie before or during the copy) - I'd argue _before_ because defect is UB already so the application was already in an undefined state and it's not worse if we skip partial copying.

Reply:
As I said, remove this warning. Fix the behavior in a followup PR if it really lacks the defect checks.

Reply:
Reworded.